### PR TITLE
Support Register Unsafely in Update

### LINKED ISSUE #3162: Remove e-mail address from registration
The subcommand `certbot update_account` allows users to update the email address in their ACME account. Currently, Certbot does not allow users to remove all email addresses from a registration. See the code for [update_account](https://github.com/certbot/certbot/blob/df1ca726f975b4f2cf96ada29785a4d35027e5f9/certbot/certbot/_internal/main.py#L684-L727).

We should allow users to do this, probably by reusing the `--register-unsafely-without-email` flag since it is usually a bad idea to have a Let's Encrypt account with no contact information.

This issue relies on https://github.com/certbot/certbot/issues/5129 being resolved which you may want to do in a PR before this one.

EDIT: My original description of this issue was very sparse and vague so I rewrote it to provide more information.

### PR DESCRIPTION
**Pull request by aeturnum**
Hopefully fixes #3162 . 

This is my first contribution to certbot so I expect I've made some mistakes, but I believe the basic changes are there.

Unsure about the type hint - one of the tox testing scripts insisted I add it, but I thought this might deviate from the type hint style used in other parts of the code.

### CODE REVIEW

**Comment by bmw** [comment]
Thanks for the PR @aeturnum. Unfortunately this doesn't work.

If you run Certbot with this PR running commands like:
```
$ sudo certbot register --email <your email> --agree-tos --no-eff-email
Saving debug log to /var/log/letsencrypt/letsencrypt.log

IMPORTANT NOTES:
 - Your account credentials have been saved in your Certbot
   configuration directory at /etc/letsencrypt. You should make a
   secure backup of this folder now. This configuration directory will
   also contain certificates and private keys obtained by Certbot so
   making regular backups of this folder is ideal.
$ sudo certbot update_account --register-unsafely-without-email
Saving debug log to /var/log/letsencrypt/letsencrypt.log
An unexpected error occurred:
TypeError: unhashable type: 'list'
Please see the logfiles in /var/log/letsencrypt for more details.
$ sudo cat /var/log/letsencrypt/letsencrypt.log
2020-08-14 19:25:37,996:DEBUG:certbot._internal.main:certbot version: 1.8.0.dev0
2020-08-14 19:25:37,996:DEBUG:certbot._internal.main:Arguments: ['--register-unsafely-without-email']
2020-08-14 19:25:37,996:DEBUG:certbot._internal.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#certbot-route53:auth,PluginEntryPoint#dns-cloudflare,PluginEntryPoint#dns-cloudxns,PluginEntryPoint#dns-digitalocean,PluginEntryPoint#dns-dnsimple,PluginEntryPoint#dns-dnsmadeeasy,PluginEntryPoint#dns-gehirn,PluginEntryPoint#dns-google,PluginEntryPoint#dns-linode,PluginEntryPoint#dns-luadns,PluginEntryPoint#dns-nsone,PluginEntryPoint#dns-ovh,PluginEntryPoint#dns-rfc2136,PluginEntryPoint#dns-route53,PluginEntryPoint#dns-sakuracloud,PluginEntryPoint#manual,PluginEntryPoint#nginx,PluginEntryPoint#null,PluginEntryPoint#standalone,PluginEntryPoint#webroot)
2020-08-14 19:25:38,026:DEBUG:certbot._internal.log:Root logging level set at 20
2020-08-14 19:25:38,027:INFO:certbot._internal.log:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2020-08-14 19:25:38,041:DEBUG:acme.client:Sending GET request to https://acme-staging-v02.api.letsencrypt.org/directory.
2020-08-14 19:25:38,043:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): acme-staging-v02.api.letsencrypt.org:443
2020-08-14 19:25:38,184:DEBUG:urllib3.connectionpool:https://acme-staging-v02.api.letsencrypt.org:443 "GET /directory HTTP/1.1" 200 724 
2020-08-14 19:25:38,185:DEBUG:acme.client:Received response:
HTTP 200 
Server: nginx
Date: Fri, 14 Aug 2020 19:25:38 GMT 
Content-Type: application/json
Content-Length: 724 
Connection: keep-alive
Cache-Control: public, max-age=0, no-cache
X-Frame-Options: DENY
Strict-Transport-Security: max-age=604800

{
  "cO4TTa_sxKk": "https://community.letsencrypt.org/t/adding-random-entries-to-the-directory/33417",
  "keyChange": "https://acme-staging-v02.api.letsencrypt.org/acme/key-change",
  "meta": {
    "caaIdentities": [
      "letsencrypt.org"
    ],  
    "termsOfService": "https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf",
    "website": "https://letsencrypt.org/docs/staging-environment/"
  },  
  "newAccount": "https://acme-staging-v02.api.letsencrypt.org/acme/new-acct",
  "newNonce": "https://acme-staging-v02.api.letsencrypt.org/acme/new-nonce",
  "newOrder": "https://acme-staging-v02.api.letsencrypt.org/acme/new-order",
  "revokeCert": "https://acme-staging-v02.api.letsencrypt.org/acme/revoke-cert"
}
2020-08-14 19:25:38,186:DEBUG:acme.client:Requesting fresh nonce
2020-08-14 19:25:38,186:DEBUG:acme.client:Sending HEAD request to https://acme-staging-v02.api.letsencrypt.org/acme/new-nonce.
2020-08-14 19:25:38,235:DEBUG:urllib3.connectionpool:https://acme-staging-v02.api.letsencrypt.org:443 "HEAD /acme/new-nonce HTTP/1.1" 200 0
2020-08-14 19:25:38,236:DEBUG:acme.client:Received response:
HTTP 200 
Server: nginx
Date: Fri, 14 Aug 2020 19:25:38 GMT 
Connection: keep-alive
Cache-Control: public, max-age=0, no-cache
Link: <https://acme-staging-v02.api.letsencrypt.org/directory>;rel="index"
Replay-Nonce: 0002CQVAODTvbghI5iiK2UgRZu-Jb_XqbmyaeCeSSNbIawk
X-Frame-Options: DENY
Strict-Transport-Security: max-age=604800


2020-08-14 19:25:38,237:DEBUG:acme.client:Storing nonce: 0002CQVAODTvbghI5iiK2UgRZu-Jb_XqbmyaeCeSSNbIawk
2020-08-14 19:25:38,237:DEBUG:certbot._internal.log:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/local/bin/certbot", line 11, in <module>
    load_entry_point('certbot', 'console_scripts', 'certbot')()
  File "/home/bmw/src/certbot/certbot/certbot/main.py", line 15, in main
    return internal_main.main(cli_args)
  File "/home/bmw/src/certbot/certbot/certbot/_internal/main.py", line 1366, in main
    return config.func(config, plugins)
  File "/home/bmw/src/certbot/certbot/certbot/_internal/main.py", line 717, in update_account
    acc.regr = cb_client.acme.update_registration(acc.regr.update(
  File "/home/bmw/src/certbot/acme/acme/client.py", line 639, in update_registration
    new_regr = self._get_v2_account(regr)
  File "/home/bmw/src/certbot/acme/acme/client.py", line 645, in _get_v2_account
    response = self._post(self.directory['newAccount'], only_existing_reg)
  File "/home/bmw/src/certbot/acme/acme/client.py", line 97, in _post
    return self.net.post(*args, **kwargs)
  File "/home/bmw/src/certbot/acme/acme/client.py", line 1201, in post
    return self._post_once(*args, **kwargs)
  File "/home/bmw/src/certbot/acme/acme/client.py", line 1211, in _post_once
    data = self._wrap_in_jws(obj, self._get_nonce(url, new_nonce_url), url, acme_version)
  File "/home/bmw/src/certbot/acme/acme/client.py", line 1017, in _wrap_in_jws
    jobj = obj.json_dumps(indent=2).encode() if obj else b'' 
  File "/home/bmw/src/certbot/venv3/lib/python3.8/site-packages/josepy/interfaces.py", line 192, in json_dumps
    return json.dumps(self, default=self.json_dump_default, **kwargs)
  File "/usr/lib/python3.8/json/__init__.py", line 234, in dumps
    return cls(
  File "/usr/lib/python3.8/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/lib/python3.8/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/home/bmw/src/certbot/venv3/lib/python3.8/site-packages/josepy/interfaces.py", line 217, in json_dump_default
    return python_object.to_partial_json()
  File "/home/bmw/src/certbot/venv3/lib/python3.8/site-packages/josepy/json_util.py", line 255, in to_partial_json
    return self.fields_to_partial_json()
  File "/home/bmw/src/certbot/venv3/lib/python3.8/site-packages/josepy/json_util.py", line 244, in fields_to_partial_json
    omitted.add((slot, value))
TypeError: unhashable type: 'list'
2020-08-14 19:25:38,241:ERROR:certbot._internal.log:An unexpected error occurred:
2020-08-14 19:25:38,242:ERROR:certbot._internal.log:TypeError: unhashable type: 'list'
```
This is at least partially because our `acme` library needs to be updated to add support for removing an email address like this before this PR can work.

I'm sorry for any confusion here but I tried to explain this at https://github.com/certbot/certbot/issues/3162.

The issue tracking the `acme` work is https://github.com/certbot/certbot/issues/5129.

**Comment by aeturnum** [comment]
Whoops! I definitely glossed over the note about #5129. I'm also embarrassed to say I didn't do my own integration testing because I wasn't confident in my ability to select appropriate arguments.

I'm confused about the error - I think my change replaces a list of strings with an empty list - I'm probably missing some conditional transformation in the bowls of `acme`.

Well then, for now I'll leave this and go look at #5129. Hopefully I'll have something for both and I'll be sure to do a full Integration test.

**Comment by aeturnum** [comment]
Oh! Hm, I didn't realize github would automagically add my changes to the pull request.

I believe this changeset will address both #3162 and #5129 . I'm nor sure about how to verify the servers' ability to receive those messages.  

The various changes are also kind of mixed up (there's no `acme` tests for instance), but I'm not sure what the project preference is on change cleanliness or separation.

File: certbot/tests/main_test.py [hunk 1]
```diff
@@ -1404,6 +1404,40 @@ def test_update_account_no_existing_accounts(self):
                  "user@example.org"])
             self.assertTrue("Could not find an existing account" in x[0])
 
+    def test_update_account_remove_email(self):
+        with mock.patch('certbot._internal.eff.prepare_subscription') as mock_prepare:
+            with mock.patch('certbot._internal.main._determine_account') as mocked_det:
+                with mock.patch('certbot._internal.main.account') as mocked_account_module:
+                    with mock.patch('certbot._internal.main.client') as mocked_client:
```
Comment (alexzorin) [inline certbot/tests/main_test.py:1411]: If mocking for the entire span of the method, [doing it like this](https://github.com/certbot/certbot/blob/860af81fef3042134b2b71a6b71b4240c75b1886/certbot/tests/main_test.py#L149-L158) can save you from excessive nesting.

Comment (aeturnum) [reply certbot/tests/main_test.py:1411]: I had used [test_update_account_with_email](https://github.com/certbot/certbot/blob/860af81fef3042134b2b71a6b71b4240c75b1886/certbot/tests/main_test.py#L1407-L1435) as a template to learn how tests were done and wasn't sure about decorator / with block preferences. Happy to refactor to use all decorators!

**Comment by aeturnum** [comment]
Not sure how to proceed here. It seems like the failed test was due to Python 3.8 failing to install tox which...seems unrelated to my changes?

Also this PR was opened focusing on #3162 and now the code addresses #5129 as well. @bmw do you have advice about how to proceed? I'm happy to split the work and re-submit separately for the two issues.

**Comment by bmw** [comment]
> It seems like the failed test was due to Python 3.8 failing to install tox which...seems unrelated to my changes?

This was a temporary issue in the Azure Pipelines image which has now been fixed and the test passes. Sorry for the trouble.

> @bmw do you have advice about how to proceed?

Thanks for the ping! Somehow I missed your commit updating `acme`. I'll try and take a look at this tomorrow.

File: certbot/certbot/_internal/main.py [hunk 2]
```diff
@@ -722,8 +722,18 @@ def update_account(config, unused_plugins):
     # so that we can also continue to use the account object with acmev1.
     acc.regr = acc.regr.update(uri=prev_regr_uri)
     account_storage.update_regr(acc, cb_client.acme)
-    eff.prepare_subscription(config, acc)
-    add_msg("Your e-mail address was updated to {0}.".format(config.email))
+
+    if config.email is None:
+        add_msg(
+            ("Your e-mail address was removed. "
+            "Make sure you backup your account key stored in {0}\n".format(
```
Comment (bmw) [inline certbot/certbot/_internal/main.py:729]: I think we should remove this sentence as I think ACME accounts are pretty disposable with the current spec.

I think the main concern with removing all of your contact information is, well, the CA cannot contact you so you won't get notifications about things like upcoming certificate expiration, revocation events, API changes, etc.

You could try to explain something like this in a CA agnostic way if you want, but I think just removing the sentence is fine since the damage is already done and this is explained in the `--register-unsafely-without-email` docs which the user had to provide.

File: acme/acme/messages.py [hunk 3]
```diff
@@ -315,7 +315,7 @@ class Registration(ResourceBody):
     # on new-reg key server ignores 'key' and populates it based on
     # JWS.signature.combined.jwk
     key = jose.Field('key', omitempty=True, decoder=jose.JWK.from_json)
-    contact = jose.Field('contact', omitempty=True, default=())
+    contact = fields.UnFalseyField('contact', omitempty=True)
```
Comment (bmw) [inline acme/acme/messages.py:318]: We're close here but there are some things I don't like about this approach. In particular, it changes the default value of the field and there are currently problems like:
```
In [3]: Registration().emails
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-48d171e81d05> in <module>
----> 1 Registration().emails

acme/acme/messages.py in emails(self)
    354     def emails(self):
    355         """All emails found in the ``contact`` field."""
--> 356         return self._filter_contact(self.email_prefix)
    357 
    358 

acme/acme/messages.py in _filter_contact(self, prefix)
    342 
    343     def _filter_contact(self, prefix):
--> 344         return tuple(
    345             detail[len(prefix):] for detail in self.contact  # pylint: disable=not-an-iterable
    346             if detail.startswith(prefix))

TypeError: 'NoneType' object is not iterable
```
We could fix the latter, but I would like to keep the default the same which will automatically fix this problem. Doing that is much harder though.

I didn't implement this so there may be problems that I'm not seeing right now, but to resolve https://github.com/certbot/certbot/issues/5129, I think we should take an approach similar to what we did in `acme/mixins.py`. I would make the following changes to `messages.UpdateRegistration`:

1. Add an `__init__` method that checks if `contact` was included in the given arguments, setting a private (prefixed with an underscore) boolean variable with the result. See `mixins.py` about how to set this attribute.
2. Add `to_partial_json` and `fields_to_partial_json` methods that call the parent method and removes the `contact` field if it wasn't set by the user and just has it's default value.

This is certainly more complex and less elegant than what was done here, however, I think it reduces the amount of unintended changes to behavior made in this PR which I think is pretty valuable.

File: certbot/certbot/_internal/main.py (hunk 2 above)
Comment (aeturnum) [reply certbot/certbot/_internal/main.py:729]: That makes sense - I got the text from the warning that `get_email()` [diplays](https://github.com/certbot/certbot/blob/master/certbot/certbot/display/ops.py#L34-L38) when you use `--register-unsafely-without-email` and I wanted to be consistent. Even the line about the email address being removed isn't a reliable description of what has happened - maybe I will change it to "Your account is no longer associated with an email address" as that will always be true.

Comment (bmw) [reply certbot/certbot/_internal/main.py:729]: Ah thanks for the link! I didn't realize that text was there. It's outdated though so I don't think we should duplicate it here.

Your change to the first sentence SGTM!

Comment (aeturnum) [reply certbot/certbot/_internal/main.py]: I realized that this will actually remove *all* contact information, including any phone numbers the user may have added. I used a version of the language that informed them of that.

**Comment by aeturnum** [comment]
Made the suggested changes. I realized that using `--register-unsafely-without-email` will also eliminate any contact telephone numbers the user may have provided. I'm not sure how to avoid this (since the regr.json on disk is purposefully blank and we can't tell if the user provided numbers), but it seems worth considering.

I'm not sure if the changelog text is appropriate but I'm happy to update it!

**Comment by aeturnum** [comment]
@bmw I think we have to keep the falsey-field with omitempty=true for the reasons outlined in my post [here](https://github.com/certbot/certbot/issues/5129#issuecomment-674565991). Basically, there are places where we save empty registration records and we either need to add empty `contact` fields to them or tell `jose` not to expect the `contact` field.

**Comment by bmw** [comment]
Oh fun. I didn't realize/had forgotten that `omitempty` had a second effect of requiring it during deserialization.  In that case, what do you think about essentially flipping the logic you added in `Registration`? Instead of removing `contact` when it wasn't explicitly set, add it when it is explicitly set, and make no changes to the `contact` field code.

It's not pretty, but again my thinking here is to minimize other behavior changes like those I flagged in https://github.com/certbot/certbot/pull/8212#discussion_r474233733.

File: certbot/CHANGELOG.md [hunk 4]
```diff
@@ -26,6 +26,8 @@ More details about these changes can be found on our GitHub repo.
   this concerns the plugin name, CLI flags, and keys in credential files.
   The prefixed form is still supported but is deprecated, and will be removed in a future release.
 * Added `--nginx-sleep-seconds` (default `1`) for environments where nginx takes a long time to reload.
+* Added the ability to remove email and phone contact information from an account 
+  using `update_account --register-unsafely-without-email`
 
 ### Changed
```
Comment (bmw) [inline certbot/CHANGELOG.md:33]: nit: There aren't many, but there are a few projects using our `acme` module as a library. Can we add a "fixed" section describing how a contact field that is explicitly set to an empty tuple is now serialized and sent to the ACME server allowing people to remove contact information from an existing registration?

File: certbot/certbot/_internal/main.py [hunk 5]
```diff
@@ -11,7 +11,7 @@
 import zope.component
 
 from acme import errors as acme_errors
-from acme.magic_typing import Union
+from acme.magic_typing import Union, Iterable
```
Comment (bmw) [inline certbot/certbot/_internal/main.py:14]: Nice catch on the `authenticator` type below! I think we should import this here as a result of that change though.
```suggestion
from acme.magic_typing import Union, Iterable, Optional
```

File: certbot/certbot/_internal/main.py [hunk 6]
```diff
@@ -703,17 +703,17 @@ def update_account(config, unused_plugins):
...
-                    "new e-mail address must\ncurrently be provided when "
-                    "updating a registration.")
+    if config.email is None and not config.register_unsafely_without_email:
         config.email = display_ops.get_email(optional=False)
 
     acc, acme = _determine_account(config)
     cb_client = client.Client(config, acc, None, None, acme=acme)
+    # Empty list of contacts in case the user is removing all emails
+
+    acc_contacts = () # type: Iterable
```
Comment (bmw) [inline certbot/certbot/_internal/main.py:713]:
```suggestion
    acc_contacts = () # type: Iterable[str]
```

File: certbot/tests/main_test.py [hunk 7]
```diff
@@ -1404,6 +1404,46 @@ def test_update_account_no_existing_accounts(self):
...
+
+            # Mock client and mock registration body to verify calls are made
+            cb_client = mock.MagicMock()
+            mock_regr_body = mock.MagicMock()
+
+            mocked_client.Client.return_value = cb_client
+            # mocked_account.regr is overwritten in update, requiring an odd mock setup
+            mocked_account.regr.body = mock_regr_body
+
+            x = self._call_no_clientmock(
```
Comment (bmw) [inline certbot/tests/main_test.py:1428]: nit: Instead of using this and mocking client here, can't we just use `self._call` and assert whatever we want on the returned mock client object?

File: certbot/tests/main_test.py [hunk 8]
```diff
@@ -1404,6 +1404,46 @@ def test_update_account_no_existing_accounts(self):
...
+                ["update_account", "--register-unsafely-without-email"])
+
+
+            # When update succeeds, the return value of register() is None
```
Comment (bmw) [inline certbot/tests/main_test.py:1432]:
```suggestion
            # When update succeeds, the return value of update_account() is None
```

File: acme/acme/messages.py [hunk 9]
```diff
@@ -340,11 +340,45 @@ def from_data(cls, phone=None, email=None, external_account_binding=None, **kwar
...
+        it should be included in serializations when it was explicitly provided.
+
+        :param jobj: Dictionary containing this Registrations' data
+        :type jobj: dict
+
+        :returns: Dictionary containing Registrations data to transmit to the server
+        :rtype: dict
+        """
+        if getattr(self, '_add_contact', False):
+            jobj['contact'] = self.contact
```
[the remark this answers is not in the document]
Comment (aeturnum) [reply acme/acme/messages.py:368]: Ah, this was a part of the data model I missed. Thank you for the catch!

File: acme/acme/messages.py [hunk 10]
```diff
@@ -340,11 +340,45 @@ def from_data(cls, phone=None, email=None, external_account_binding=None, **kwar
 
         return cls(**kwargs)
 
+    def __init__(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (aeturnum) [reply acme/acme/messages.py:343]: I think I can do that, but I don't quite follow what you mean. The code doesn't currently set contact, exactly, but sets a private variable about the presence of contact. I can certainly expand the existing brief comment on line 344 to explain the overall thinking, but I want to make sure I'm not misunderstanding where you'd like to see a comment....

[the remark this answers is not in the document]
Comment (bmw) [reply acme/acme/messages.py:343]: Sorry. GitHub won't let me comment on where I was referring to because you didn't modify the code there.

I was proposing we add a comment above [the definition of the contact field](https://github.com/certbot/certbot/blob/2a047eb526ee02d4bb3bb3668260ac6d481b99d3/acme/acme/messages.py#L318). My thinking is just to flag that the field no longer works as a normal field with the value of `omitempty=True` where the field is defined in case people miss your code here.

File: certbot/certbot/_internal/main.py (hunk 5 above)
Comment (aeturnum) [reply certbot/certbot/_internal/main.py:14]: I thought I had tried this when I added Iterable, and it turns out I did try. For reasons I don't understand, this fails the `lint` target (even though `Optional` is obviously used on line 593 in a type description):
`************* Module certbot._internal.main
certbot/certbot/_internal/main.py:14:0: W0611: Unused Optional imported from acme.magic_typing (unused-import)`

**Comment by aeturnum** [comment]
Resolved everything except for the addition of an Optional import, with details as to why on the appropriate review.

File: acme/acme/messages.py [hunk 11]
```diff
@@ -328,12 +331,22 @@ class Registration(ResourceBody):
...
+        details_provided = 'contact' in kwargs and kwargs['contact'] is not None
+
         details = list(kwargs.pop('contact', ()))
         if phone is not None:
             details.append(cls.phone_prefix + phone)
         if email is not None:
             details.extend([cls.email_prefix + mail for mail in email.split(',')])
-        kwargs['contact'] = tuple(details)
+
+        # Serialization behavior change based on providing `contacts` to the constructor
```
Comment (bmw) [inline acme/acme/messages.py:343]:
```suggestion
        # Serialization behavior changes based on providing `contacts` to the constructor
```

File: acme/acme/messages.py [hunk 12]
```diff
@@ -328,12 +331,22 @@ class Registration(ResourceBody):
     @classmethod
     def from_data(cls, phone=None, email=None, external_account_binding=None, **kwargs):
         """Create registration resource from contact details."""
+        # extra variable to detect non-null empty lists or tuples
+        details_provided = 'contact' in kwargs and kwargs['contact'] is not None
```
Comment (bmw) [inline acme/acme/messages.py:335]: I don't `contact` should ever be `None`. If it is and `phone` or `email` is provided, we'll call `None.append` which won't go well.
```suggestion
        details_provided = 'contact' in kwargs
```

File: acme/acme/messages.py [hunk 13]
```diff
@@ -328,12 +331,22 @@ class Registration(ResourceBody):
...
         if phone is not None:
             details.append(cls.phone_prefix + phone)
         if email is not None:
             details.extend([cls.email_prefix + mail for mail in email.split(',')])
-        kwargs['contact'] = tuple(details)
+
+        # Serialization behavior change based on providing `contacts` to the constructor
+        # So we add the `contact` argument if:
+        # - `phone` or `email` or `contact` `kwargs` are passed with values
+        # - a empty list or tuple is provided in the `kwargs`
```
Comment (bmw) [inline acme/acme/messages.py:346]: I think we should remove or clarify this line. An empty list or tuple is provided in `kwargs` for what?

File: acme/acme/messages.py [hunk 14]
```diff
@@ -342,7 +355,7 @@ def from_data(cls, phone=None, email=None, external_account_binding=None, **kwar
 
     def __init__(self, **kwargs):
         """Note if the user provides a value for the `contact` member."""
-        if 'contact' in kwargs:
+        if 'contact' in kwargs and kwargs['contact'] is not None:
```
Comment (bmw) [inline acme/acme/messages.py:358]: Same comment about how I don't think `contact` should ever be done.
```suggestion
        if 'contact' in kwargs:
```

File: certbot/certbot/_internal/main.py (hunk 5 above)
Comment (bmw) [reply certbot/certbot/_internal/main.py:14]: In that case I think this line should be:
```suggestion
from acme.magic_typing import Union, Iterable, Optional  # pylint: disable=unused-import
```
This appears to be a `pylint` bug.

The reason I think it's good to have is I don't entirely understand how we're currently getting away without importing here and I worry it could break. For instance, `mypy` may be relying on `Optional` having been imported in another module that's checked before this one and if that file changes, we may suddenly start getting errors here.

File: certbot/tests/main_test.py [hunk 15]
```diff
@@ -1404,6 +1404,44 @@ def test_update_account_no_existing_accounts(self):
...
+    def test_update_account_remove_email(self, mocked_account_module, mock_prepare, mock_det_acc):
+        # Mock account storage and the account object returned
+        mocked_storage = mock.MagicMock()
+        mocked_account = mock.MagicMock()
+
+        mocked_account_module.AccountFileStorage.return_value = mocked_storage
+        mocked_storage.find_all.return_value = [mocked_account]
+        mock_det_acc.return_value = (mocked_account, "foo")
+
+        # Mock client and mock registration body to verify calls are made
```
Comment (bmw) [inline certbot/tests/main_test.py:1419]:
```suggestion
        # Mock registration body to verify calls are made
```

File: certbot/tests/main_test.py [hunk 16]
```diff
@@ -1404,6 +1404,44 @@ def test_update_account_no_existing_accounts(self):
...
+            ["update_account", "--register-unsafely-without-email"])
+
+
+        # When update succeeds, the return value of update_account() is None
+        self.assertTrue(x[0] is None)
+        # and we got supposedly did update the registration from
+        # the server
+        client_mock = x[3]
+
+        self.assertEqual(client_mock.mock_calls[1][0], 'Client().acme.update_registration')
```
Comment (bmw) [inline certbot/tests/main_test.py:1435]: I think this is relying on the exact order calls were made on the client object which seems fragile to me. Instead, what about something like:
```
        mock_update_reg = client_mock.Client().acme.update_registration
        self.assertTrue(mock_update_reg.called)
```

File: acme/acme/messages.py (hunk 12 above)
Comment (aeturnum) [reply acme/acme/messages.py:335]: I don't think that's right, because we provide a default to `kwargs.pop` on line 337 (and did in the previous version of the code), so we don't risk trying to call methods on None.

File: acme/acme/messages.py (hunk 13 above)
Comment (aeturnum) [reply acme/acme/messages.py:346]: So if the user passes `contact=[]`, we need to replace the `list` with a `tuple` (`list` is not hashable and will throw an exception). This was previously handled by the old code path because the [setting of the `contact` field of kwargs was unconditional](https://github.com/certbot/certbot/blob/master/acme/acme/messages.py#L336) (because explicitly providing an argument didn't change behavior).

However, if the user passes `contact=None`, we need to not set `contact,` because (I didn't actually test this so maybe it's ok?) contact=None will probably save a `null` in the `json` (creating a potential de-serialization problem).

File: acme/acme/messages.py (hunk 12 above)
Comment (bmw) [reply acme/acme/messages.py:335]: Well I was wrong about the place we blow up:
```
In [1]: kwargs = {'contact': None}

In [2]: details = list(kwargs.pop('contact', ()))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-b877192e83a3> in <module>
----> 1 details = list(kwargs.pop('contact', ()))

TypeError: 'NoneType' object is not iterable
```
`kwargs.pop('contact', ())` will return `None` if `contact` is in `kwargs` and set to `None`. It will only return an empty tuple if it is not provided at all.

File: acme/acme/messages.py (hunk 14 above)
Comment (aeturnum) [reply acme/acme/messages.py:358]: As I said above, I am concerned about what happens in this new code if `None` is explicitly provided, leading to a situation where de-serialization results in `contact` containing `None` instead of a `tuple` and crashing when we try to iterate over it.

File: certbot/tests/main_test.py (hunk 16 above)
Comment (aeturnum) [reply certbot/tests/main_test.py:1435]: Ah thank you! The python documentation on mocks elides practical usage and I did not realize I could include the `Client()` in the call chain. That's much better and I appreciate the suggestion.

File: acme/acme/messages.py (hunk 12 above)
Comment (aeturnum) [reply acme/acme/messages.py:335]: Oops, I forgot that defaults are used when the key isn't defined, not when the key is `None`. Some experienced python dev I am 😳

File: acme/acme/messages.py (hunk 14 above)
Comment (aeturnum) [reply acme/acme/messages.py:358]: Ah, nope, this was all chasing imaginary ghosts. Thank you for the catch.

**Comment by aeturnum** [comment]
Oops - @bmw I just saw that I added my changelog messages to the wrong section (1.7.0 instead of 1.8.0). Should I open a new PR to move them to the proper place?

**Comment by bmw** [comment]
Yes please! Thanks for catching that.
